freetype/__init__.py

```diff
@@ -975,22 +976,39 @@ def __init__( self, filename, index = 0 ):
         library = get_handle( )
         face = FT_Face( )
         self._FT_Face = None
-        #error = FT_New_Face( library, filename, 0, byref(face) )
+        #error = FT_New_Face( library, path_or_stream, 0, byref(face) )
         self._filebodys = []
-        try:
-            u_filename = c_char_p(_encode_filename(filename))
-            error = FT_New_Face( library, u_filename, index, byref(face) )
-        except UnicodeError:
-            with open(filename, mode='rb') as f:
-                filebody = f.read()
-            error = FT_New_Memory_Face( library, filebody, len(filebody),
-                                        index, byref(face) )
-            self._filebodys.append(filebody)  # prevent gc
-        if error: raise FT_Exception( error )
-        self._filename = filename
+        if hasattr(path_or_stream, "read"):
+            error = self._init_from_memory(library, face, index, path_or_stream.read())
+        else:
+            try:
+                error = self._init_from_file(library, face, index, path_or_stream)
+            except UnicodeError:
+                with open(path_or_stream, mode="rb") as f:
+                    filebody = f.read()
+                error = self._init_from_memory(library, face, index, filebody)
+        if error:
+            raise FT_Exception(error)
+        self._filename = path_or_stream
         self._index = index
         self._FT_Face = face
 
+    def _init_from_file(self, library, face, index, path):
+        u_filename = c_char_p(_encode_filename(path))
+        error = FT_New_Face(library, u_filename, index, byref(face))
+        return error
+
+    def _init_from_memory(self, library, face, index, byte_stream):
+        error = FT_New_Memory_Face(
+            library, byte_stream, len(byte_stream), index, byref(face)
+        )
+        self._filebodys.append(byte_stream)  # prevent gc
+        return error
+
+    @classmethod
+    def from_memory(cls, bytes_, index=0):
+         return cls(io.BytesIO(bytes_), index)
```

Comment:
Why not call _init_from_memory directly here, instead of making a temp BytesIO?

Reply:
Also, frombytes (or if you like underscores from_bytes) reads better to me, but that’s just taste

Reply:
It's a `@classmethod` and I have to go through `__init__` somehow, no?

Reply:
not necessarily. You can call `self = cls.__new__(cls)` to allocate the new object and then initialize it

Reply:
basically, calling `__new__` explicitly in the alternative constructor would circumvent `__init__`

Reply:
never mind, it's also fine like that. It doesn't add too much overhead to create the BytesIO (it's just a thin wrapper around some bytes string), and the whole from_bytes classmethod is mostly syntactic sugar, sparing the user an extra `from io import BytesIO` basically.

Reply:
Hm, I'd need to replicate the init logic then, no? This is something for a refactoring with breaking changes...